File: union-find/java/src/QuickFind.java
/**
 * Created by petergoldsborough on 08/12/15.
 */

public class QuickFind extends UnionFind
{
    QuickFind(int N)
    {
        super(N);
    }

    public void union(int p, int q)
    {
        if (! connected(p, q))
        {
            int p_id = _id[p];

            int q_id = _id[q];

            for (int i = 0; i < _N; ++i)
            {
                if (_id[i] == p_id) _id[i] = q_id;
            }
        }
    }

    public boolean connected(int p, int q)
    {
        return _id[p] == _id[q];
    }
};

File: union-find/python/QuickFind.py
#!/usr/local/bin/python3
# -*- coding: utf-8 -*-

import UnionFind

class QuickFind(UnionFind.UnionFind):

	def __init__(self, N):
		super().__init__(N)
		self.id = [x for x in range(N)]

	def union(self, p, q):
		p_id = self.id[p]
		q_id = self.id[q]

		if p_id != q_id:
			for n,i in enumerate(self.id):
				if i == p_id:
					self.id[n] = q_id

	def connected(self, p, q):
		return self.id[p] == self.id[q]

	def __repr__(self):
		return str(self.id)

def main():
	UnionFind.main(QuickFind)

if __name__ == "__main__":
	main()



